satviz-parsers/src/main/java/edu/kit/satviz/parsers/ClauseFile.java

```diff
@@ -0,0 +1,30 @@
+package edu.kit.satviz.parsers;
+
+import edu.kit.satviz.sat.ClauseUpdate;
+import java.io.InputStream;
+import java.util.Scanner;
+
+/**
+ * This class is used to parse an <code>InputStream</code> iteratively into
+ * <code>ClauseUpdates</code> by implementing the <code>Iterable</code> interface.
+ */
+public abstract class ClauseFile implements Iterable<ClauseUpdate>, AutoCloseable {
+
+  protected final Scanner scanner;
+
+  protected ClauseFile(InputStream in) {
```

Comment:
Warum übergeben wir hier im Konstruktor einen InputStream? Mit einem InputStream können wir die Datei nur ein mal lesen, und damit auch nur einen einzigen Iterator erstellen.
Hier sollte ein Argument übergeben werden, von dem man beliebig viele unabhängige Streams oder Scanner erzeugen kann (z.B. File).

Reply:
`File` ist beschränkt auf Dateien im Dateiensystem, dabei soll der input aus beliebiger Quelle kommen können. Also wenn man einen "mehrmals lesen"-Mechanismus einbauen möchte, sollte man das anders lösen.

Reply:
Da wir intern einen Scanner benutzen, sind unsere Möglichkeiten sowieso beschränkt. Die einzigen Objekte aus denen man mehrere unabhängige Scanner machen kann, sind File und String. Ich schlage vor wir schreiben für die beiden jeweils einen Konstruktor.

Reply:
Wir können das auch vorerst ignorieren, wenn es dringendere Sachen gibt. Aber das Wunschverhalten ist dass iterator() immer einen neuen Iterator zurückgibt.

Reply:
Davon bin ich persönlich kein Fan, damit entfernt man nämlich fast genauso viel Flexibilität wie wenn man nur `File`s erlaubt. Aber selbst wenn der `InputStream`-Konstruktor bleibt, hat man dann inkonsistentes Verhalten und special casing je nachdem, wie die Instanz erstellt wurde.
Die einzige "gute" Lösung wäre ein Cache. Aber ich sehe es erstmal überhaupt nicht als notwendig, hier mehrfache Iteration zu erlauben. Es ist weder vom `Iterable` Interface gefordert noch brauchen wir es irgendwo in unserer Anwendung.

Reply:
Das ist für mich schon das erwartete Verhalten (Iterator soll das Iterable nicht unerwartet verändern), auch wenn es nicht explizit gefordert wird. Aber du hast recht, streng genommen brauchen wir das nicht.